crates/primitives/src/signature.rs

```diff
@@ -0,0 +1,855 @@
+    ///
+    /// [`Address`]: crate::Address
+    #[inline]
+    pub fn recover_address_from_msg<T: AsRef<[u8]>>(
+        &self,
+        msg: T,
+    ) -> Result<crate::Address, SignatureError> {
+        self.recover_from_msg(msg).map(|vk| crate::Address::from_public_key(&vk))
+    }
+
+    /// Recovers an [`Address`] from this signature and the given prehashed message.
+    ///
+    /// [`Address`]: crate::Address
+    #[inline]
+    pub fn recover_address_from_prehash(
+        &self,
+        prehash: crate::B256,
+    ) -> Result<crate::Address, SignatureError> {
+        self.recover_from_prehash(prehash).map(|vk| crate::Address::from_public_key(&vk))
+    }
+
+    /// Recovers a [`VerifyingKey`] from this signature and the given message by first prefixing and
+    /// hashing the message according to [EIP-191](crate::eip191_hash_message).
+    ///
+    /// [`VerifyingKey`]: k256::ecdsa::VerifyingKey
+    #[inline]
+    pub fn recover_from_msg<T: AsRef<[u8]>>(
+        &self,
+        msg: T,
+    ) -> Result<k256::ecdsa::VerifyingKey, SignatureError> {
+        self.recover_from_prehash(crate::eip191_hash_message(msg))
+    }
+
+    /// Recovers a [`VerifyingKey`] from this signature and the given prehashed message.
+    ///
+    /// [`VerifyingKey`]: k256::ecdsa::VerifyingKey
+    #[inline]
+    pub fn recover_from_prehash(
+        &self,
+        prehash: crate::B256,
+    ) -> Result<k256::ecdsa::VerifyingKey, SignatureError> {
+        let this = self.normalize_s().unwrap_or(*self);
+        k256::ecdsa::VerifyingKey::recover_from_prehash(
+            prehash.as_slice(),
+            &this.inner,
+            this.recid(),
+        )
+        .map_err(Into::into)
+    }
+
+    #[cfg(feature = "rlp")]
+    pub fn decode_rlp_vrs(buf: &mut &[u8]) -> Result<Self, alloy_rlp::Error> {
+        use alloy_rlp::Decodable;
+
+        let parity: u64 = Decodable::decode(buf)?;
+        let r = Decodable::decode(buf)?;
+        let s = Decodable::decode(buf)?;
+
+        Self::from_rs_and_parity(r, s, parity)
+            .map_err(|_| alloy_rlp::Error::Custom("attempted to decode invalid field element"))
+    }
+}
+
+impl Signature<()> {
+    /// Parses a signature from a byte slice, with a v value
+    ///
+    /// # Panics
+    ///
+    /// If the slice is not at least 64 bytes long.
+    #[inline]
+    pub fn from_bytes_and_parity<T: Into<Parity>>(
+        bytes: &[u8],
+        parity: T,
+    ) -> Result<Self, SignatureError> {
+        let r = U256::from_be_slice(&bytes[..32]);
+        let s = U256::from_be_slice(&bytes[32..64]);
+        Self::from_rs_and_parity(r, s, parity)
+    }
+
+    /// Instantiate from v, r, s.
+    pub fn from_rs_and_parity<T: Into<Parity>>(
+        r: U256,
+        s: U256,
+        parity: T,
+    ) -> Result<Self, SignatureError> {
+        Ok(Self { inner: (), v: parity.into(), r, s })
+    }
+}
+
+impl<S: Copy + Eq> Signature<S> {
```

Comment:
don't think this whole block needs Copy bound, it's ok since both generics we use do implement it tho